spring-integration-core/src/main/java/org/springframework/integration/support/MutableMessageBuilder.java

```diff
@@ -73,6 +78,10 @@ public T getPayload() {
 	 */
 	public static <T> MutableMessageBuilder<T> fromMessage(Message<T> message) {
 		Assert.notNull(message, "message must not be null");
+		if (message instanceof CloneableMessage) {
+			logger.warn("For CloneableMessages consider to use regular MessageBuilder. " +
+					"Otherwise custom Message context may be lost.");
+		}
```

Comment:
As I said, I am ok with a simple fix for 4.3, but we need a generic solution for 5.0 - but in any case, the dispatcher should not have to be aware of nested inner classes in some subset of channel interceptors.

Hence we need the interface, but I'm ok if you want to put the conditional build logic in the dispatcher - and we can revisit the general builder solution if/when another use case arises.

i.e. the middle ground (for 4.3 and 5.0) is your original solution but with an interface rather than a direct reference to the interceptor inner class.